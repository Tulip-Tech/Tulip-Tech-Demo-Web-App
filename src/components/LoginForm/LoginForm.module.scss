@use 'src/styles/variables';
@use 'src/styles/theme';

.title {
  margin-bottom: 24px;
  font-weight: var(--body-font-weight-bold);
  font-size: 24px;
}

.link {
  margin-bottom: 24px;
}

.bottom {
  margin-top: 24px;
  text-align: center;
}
