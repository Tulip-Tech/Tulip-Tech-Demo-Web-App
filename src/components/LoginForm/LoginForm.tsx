import React from 'react';
import { useLocation } from 'react-router';
import { useTranslation } from 'react-i18next';

import styles from './LoginForm.module.scss';

import useToggle from '#src/hooks/useToggle';
import TextField from '#components/TextField/TextField';
import Button from '#components/Button/Button';
import Link from '#components/Link/Link';
import IconButton from '#components/IconButton/IconButton';
import Visibility from '#src/icons/Visibility';
import VisibilityOff from '#src/icons/VisibilityOff';
import FormFeedback from '#components/FormFeedback/FormFeedback';
import LoadingOverlay from '#components/LoadingOverlay/LoadingOverlay';
import { testId } from '#src/utils/common';
import { addQueryParam } from '#src/utils/location';
import type { FormErrors } from '#types/form';
import type { LoginFormData } from '#types/account';

type Props = {
  onSubmit: React.FormEventHandler<HTMLFormElement>;
  onChange: React.ChangeEventHandler<HTMLInputElement | HTMLTextAreaElement>;
  error?: string;
  errors: FormErrors<LoginFormData>;
  values: LoginFormData;
  submitting: boolean;
  siteName?: string;
};

const LoginForm: React.FC<Props> = ({ onSubmit, onChange, values, errors, submitting, siteName }: Props) => {
  const [viewPassword, toggleViewPassword] = useToggle();
  const { t } = useTranslation('account');
  const location = useLocation();

  return (
    <form onSubmit={onSubmit} data-testid={testId('login-form')} noValidate>
      <h2 className={styles.title}>{t('login.sign_in')}</h2>
      {errors.form ? <FormFeedback variant="error">{errors.form}</FormFeedback> : null}
      <TextField
        value={values.email}
        onChange={onChange}
        label={t('login.email')}
        placeholder={t('login.email')}
        error={!!errors.email || !!errors.form}
        helperText={errors.email}
        name="email"
        type="email"
        required
        testId="login-email-input"
      />
      <TextField
        value={values.password}
        onChange={onChange}
        label={t('login.password')}
        placeholder={t('login.password')}
        error={!!errors.password || !!errors.form}
        helperText={errors.password}
        name="password"
        type={viewPassword ? 'text' : 'password'}
        rightControl={
          <IconButton aria-label={viewPassword ? t('login.hide_password') : t('login.view_password')} onClick={() => toggleViewPassword()}>
            {viewPassword ? <Visibility /> : <VisibilityOff />}
          </IconButton>
        }
        required
        testId="login-password-input"
      />
      {submitting && <LoadingOverlay transparentBackground inline />}
      <Link className={styles.link} to={addQueryParam(location, 'u', 'forgot-password')}>
        {t('login.forgot_password')}
      </Link>
      <Button type="submit" label={t('login.sign_in')} variant="contained" color="primary" size="large" disabled={submitting} fullWidth />
      <p className={styles.bottom}>
        {t('login.not_registered', { siteName })} <Link to={addQueryParam(location, 'u', 'create-account')}>{t('login.sign_up')}</Link>
      </p>
    </form>
  );
};

export default LoginForm;
