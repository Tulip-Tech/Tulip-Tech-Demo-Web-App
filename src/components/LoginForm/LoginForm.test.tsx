import React from 'react';
import { fireEvent } from '@testing-library/react';

import LoginForm from './LoginForm';

import { renderWithRouter } from '#test/testUtils';

describe('<LoginForm>', () => {
  test('renders and matches snapshot', () => {
    const { container } = renderWithRouter(
      <LoginForm
        onSubmit={vi.fn()}
        onChange={vi.fn()}
        values={{
          email: '',
          password: '',
        }}
        errors={{}}
        submitting={false}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test('sets the correct values in the form fields', () => {
    const { getByLabelText } = renderWithRouter(
      <LoginForm
        onSubmit={vi.fn()}
        onChange={vi.fn()}
        values={{
          email: 'myemail@email.com',
          password: 'mypassword',
        }}
        errors={{}}
        submitting={false}
      />,
    );

    expect(getByLabelText('login.email')).toHaveValue('myemail@email.com');
    expect(getByLabelText('login.password')).toHaveValue('mypassword');
  });

  test('sets the correct errors in the form fields', () => {
    const { queryByText } = renderWithRouter(
      <LoginForm
        onSubmit={vi.fn()}
        onChange={vi.fn()}
        values={{
          email: 'myemail@email.com',
          password: 'mypassword',
        }}
        errors={{ email: 'Email error', password: 'Password error', form: 'Form error' }}
        submitting={false}
      />,
    );

    expect(queryByText('Email error')).toBeDefined();
    expect(queryByText('Password error')).toBeDefined();
    expect(queryByText('Form error')).toBeDefined();
  });

  test('disables the submit button when submitting is true', () => {
    const { getByRole } = renderWithRouter(
      <LoginForm
        onSubmit={vi.fn()}
        onChange={vi.fn()}
        values={{
          email: 'myemail@email.com',
          password: 'mypassword',
        }}
        errors={{}}
        submitting={true}
      />,
    );

    expect(getByRole('button', { name: 'login.sign_in' })).toBeDisabled();
  });

  test('calls the onSubmit callback when the form gets submitted', () => {
    const onSubmit = vi.fn();
    const { getByTestId } = renderWithRouter(
      <LoginForm
        onSubmit={onSubmit}
        onChange={vi.fn()}
        values={{
          email: 'myemail@email.com',
          password: 'mypassword',
        }}
        errors={{}}
        submitting={true}
      />,
    );

    fireEvent.submit(getByTestId('login-form'));

    expect(onSubmit).toBeCalled();
  });

  test('calls the onChange callback when a text field changes', () => {
    const onChange = vi.fn();
    const { getByLabelText } = renderWithRouter(
      <LoginForm
        onSubmit={vi.fn()}
        onChange={onChange}
        values={{
          email: '',
          password: '',
        }}
        errors={{}}
        submitting={true}
      />,
    );

    fireEvent.change(getByLabelText('login.email'), { target: { value: 'email' } });
    fireEvent.change(getByLabelText('login.password'), { target: { value: 'password' } });

    expect(onChange).toBeCalledTimes(2);
  });
});
