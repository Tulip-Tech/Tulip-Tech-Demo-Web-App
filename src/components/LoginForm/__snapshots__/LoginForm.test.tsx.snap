// Vitest Snapshot v1

exports[`<LoginForm> > renders and matches snapshot 1`] = `
<div>
  <form
    data-testid="login-form"
    novalidate=""
  >
    <h2
      class="_title_7bd5f4"
    >
      login.sign_in
    </h2>
    <div
      class="_textField_e16c1b"
      data-testid="login-email-input"
    >
      <label
        class="_label_e16c1b"
        for="text-field_1235_email"
      >
        login.email
      </label>
      <div
        class="_container_e16c1b"
      >
        <input
          class="_input_e16c1b"
          id="text-field_1235_email"
          name="email"
          placeholder="login.email"
          required=""
          type="email"
          value=""
        />
      </div>
    </div>
    <div
      class="_textField_e16c1b _rightControl_e16c1b"
      data-testid="login-password-input"
    >
      <label
        class="_label_e16c1b"
        for="text-field_1235_password"
      >
        login.password
      </label>
      <div
        class="_container_e16c1b"
      >
        <input
          class="_input_e16c1b"
          id="text-field_1235_password"
          name="password"
          placeholder="login.password"
          required=""
          type="password"
          value=""
        />
        <div
          class="_control_e16c1b"
        >
          <div
            aria-label="login.view_password"
            class="_iconButton_0fef65"
            role="button"
            tabindex="0"
          >
            <svg
              aria-hidden="true"
              class="_icon_1e9999"
              focusable="false"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="M11.83,9L15,12.16C15,12.11 15,12.05 15,12A3,3 0 0,0 12,9C11.94,9 11.89,9 11.83,9M7.53,9.8L9.08,11.35C9.03,11.56 9,11.77 9,12A3,3 0 0,0 12,15C12.22,15 12.44,14.97 12.65,14.92L14.2,16.47C13.53,16.8 12.79,17 12,17A5,5 0 0,1 7,12C7,11.21 7.2,10.47 7.53,9.8M2,4.27L4.28,6.55L4.73,7C3.08,8.3 1.78,10 1,12C2.73,16.39 7,19.5 12,19.5C13.55,19.5 15.03,19.2 16.38,18.66L16.81,19.08L19.73,22L21,20.73L3.27,3M12,7A5,5 0 0,1 17,12C17,12.64 16.87,13.26 16.64,13.82L19.57,16.75C21.07,15.5 22.27,13.86 23,12C21.27,7.61 17,4.5 12,4.5C10.6,4.5 9.26,4.75 8,5.2L10.17,7.35C10.74,7.13 11.35,7 12,7Z"
                />
              </g>
            </svg>
          </div>
        </div>
      </div>
    </div>
    <a
      class="_link_a14c64 _link_7bd5f4"
      href="/?u=forgot-password"
    >
      login.forgot_password
    </a>
    <button
      aria-disabled="false"
      class="_button_f8f296 _primary_f8f296 _fullWidth_f8f296 _large_f8f296"
      type="submit"
    >
      <span>
        login.sign_in
      </span>
    </button>
    <p
      class="_bottom_7bd5f4"
    >
      login.not_registered
       
      <a
        class="_link_a14c64"
        href="/?u=create-account"
      >
        login.sign_up
      </a>
    </p>
  </form>
</div>
`;
